### ISSUE TITLE
[binary-search]: Current implementation is much less efficient than a linear search

### ISSUE THREAD

**Issue by NobbZ** [exercism/xelixir#176]
The conclusion in the subject is from a very brief look at the code.

In general, you can't search a linked list, in a way that is more efficient than a linear search, due to its nature.

To access the element in the middle, you have to iterate the first half of the list completely, while dooing this recursively, you might even read over the searched element multiple times.

Also when splitting the list you do again fly over the first half (even doing a copy when the element is there!)

I already had a similar discussion in a project which aims on implementing some prominent CS datastructures and algorithms over it, perhaps you want to take a read at sashaafm/exads#14 and then deprecate/remove this exercise from this track.

**Comment by bernardoamc** [comment]
You are right, just saw the `fetch` implementation and it is indeed a linear scan. How to proceed so people will not be able to implement this exercise in this track again?

**Comment by NobbZ** [comment]
IIRC there are keys in the JSON which mark exercises as deprecated when trying to move them out of the track or to mark them as not suitable for a track. But I don't remember their exact names. perhaps @kytrinyx can help out here?

**Comment by NobbZ** [comment]
@kytrinyx as we are discussing it here right now, how was the link to check for other tracks having the exercise?  Then I could take a looking there are others around that implement the binary search on an inappropriate datastructure and perhaps file another issue over at x-common if necessary.

**Comment by kytrinyx** [comment]
> IIRC there are keys in the JSON which mark exercises as deprecated when trying to move them out of the track or to mark them as not suitable for a track.

In the config.json there is a key `deprecated` (array of problem slugs) for problems that are already implemented but should not be delivered via `fetch`, and `foregone` for problems that we want to avoid implementing.

> how was the link to check for other tracks having the exercise?

If you look at the output of http://x.exercism.io/v3/problems/binary-search/todo look under the key 'implementations' which lists the links to the directory of the implementation on GitHub.

**Comment by devonestes** [comment]
Can I ask a stupid question - why is the efficiency of the implementation important for our purposes? Personally, I don't see a huge problem in there being some less-than-optimal efficiency in the code that's use here. As long as it's part of a valuable learning experience I think it's just fine.

**Comment by NobbZ** [comment]
to require the student to solve the binary search over a linked list does
imply that it would be in general efficient and the student might think it
were better than a linear search because it's written in the readme.

Devon Estes <notifications@github.com> schrieb am Fr., 5. Aug. 2016 00:22:

**Comment by devonestes** [comment]
Ok, that makes sense. We certainly don't want to give anyone incorrect information! That sounds like it might be solvable in the README, or maybe in a comment in the solution file since it's a rather language-specific caveat. How about something like this:

```elixir
  @spec search(Enumerable.t, integer) :: {:ok, integer} | :not_found
  def search(list, key) do
    # NOTE: Lists in Elixir are implemented as a Linked List, which means using binary search is
    #       actually less efficient than linear search. Please, don't do this in production code!
    #       Now that you know that, have fun solving this problem!
  end
```

If we put it right in the body of the function that users are expected to write, I think that gives us a very high likelihood that they'll read it.

Even if this isn't the most efficient algorithm to solve this problem, in this language, with this data type, it's still a good concept to know, so I think keeping it makes sense. Plus, then, if folks didn't know about this implementation detail of the language before this exercise, they should know about it after, which is kind of valuable! That's just my two cents, though - I'm happy to go either way.

**Comment by lpil** [comment]
Alternatively you could use arrays, which do exist in Elixir/Erlang. See the `:array` module.

Their use is not at all idiomatic, but I think it's far better than misleading people by using lists like this.

**Comment by tejasbubane** [comment]
> If we put it right in the body of the function that users are expected to write, I think that gives us a very high likelihood that they'll read it.

Although putting it in solution file has the most likelihood of people reading it, I think this should go in `README`. This is because people tend to keep whatever is given in the stub file as is when they submit the solution. We can put a `HINTS.md` file which will get appended to `README.md` when users fetch the exercise.

**Comment by lpil** [comment]
I think people rarely read the documents, so this will continue to be an exercise that teaches extremely bad habits.

**Comment by rbasso** [comment]
> Although putting it in solution file has the most likelihood of people reading it, I think this should go in README. This is because people tend to keep whatever is given in the stub file as is when they submit

I can confirm that. In the Haskell track, we added instructions as comments in the stub solutions, but people insisted on submitting solutions maintaining the comments, as if they where sacred.

Now we started to use `HINTS.md`, but I still cannot evaluate the effectiveness of this new approach.

**Comment by kytrinyx** [comment]
> I can confirm that. In the Haskell track, we added instructions as comments in the stub solutions, but most people insisted on submitting solutions maintaining the comments, as if they where sacred.

Same in the Go track, even though we explicitly asked them to delete the comments. I implemented a bot that gives them feedback asking them to remove stub comments.

> this will continue to be an exercise that teaches extremely bad habits.

I think we should avoid this at all costs—if we can't figure out how to turn the exercise into something that teaches good habits I would recommend deprecating it.

**Comment by lpil** [comment]
We can use the array module but this would involve using a less common Erlang module and reading of the Erlang docs, so we may want to push this further back in the course.

It's still quite an odd thing to do in Elixir though.

**Comment by martinsvalin** [comment]
Is it cheating to convert the list to a balanced tree, and use `:gb_trees.lookup`? 😁 

You'll get a tree from

    tree = list
      |> Enum.with_index
      |> Enum.map(fn {el, i} -> {i, el} end)
      |> :gb_trees.from_orddict

Then simply

    case :gb_trees.lookup(needle, tree) do
      {:value, index} -> {:ok, index}
      :none -> :not_found

I'm just trying to point out that it's perfectly possibly to do a nice binary search in Elixir/Erlang without leaving the stdlib. But you'll have to convert your data into something more appropriate than linked lists.

Anyway, to drive home the actual exercise of implementing binary search, but using a data structure with constant time lookup, how about tuples as input? It'll be awkward to check that it's sorted, so maybe drop the ArgumentError. It's not essential to the algorithm.

    defmodule BinarySearch do
      def search({}, _key), do: :not_found
      def search(tuple, key) do
        do_search(tuple, key, 0, tuple_size(tuple))
      end

      defp do_search(tuple, key, low, high) when high < low, do: :not_found
      defp do_search(tuple, key, low, high) do
        middle = div(high + low, 2)
        case elem(tuple, middle) do
          ^key -> {:ok, middle}
          mid_value when key < mid_value -> do_search(tuple, key, low, middle - 1)
          mid_value when key > mid_value -> do_search(tuple, key, middle + 1, high)
        end
      end
    end

Should be O(log n). Tuples are expensive to update, but we never do in this exercise. It's a reasonable choice. I did a trivial update to the test suite (dropped argument error, wrapped elements in curlies instead of brackets). Tests are green. This isn't a particularly invasive change to the exercise.

**Comment by lpil** [comment]
I think that makes a lot more sense than using lists. :)

**Comment by martinsvalin** [comment]
I can PR my changes to the exercise (tuple inputs), but I'll wait for more feedback.

**Comment by parkerl** [comment]
I'm sold.

**Comment by NobbZ** [comment]
Converting the list to something is an O(n) operation in itself, searching the converted stuff adds on top, so we will always be less efficient than using a linear search over the list.

Also I do not like the tuple thing. Tuples are not supposed to be used like this. It would teach unidiomiatic stuff as well.

So I'd like to propose the following:

Let's use erlang `arrays` module for this and make clear in the stub file or HINTS.md, that this exercise is meant to teach interfacing with erlang modules and types ;) And also that usually `arrays` is not idiomatic Elixir.

**Comment by lpil** [comment]
Constructing tuples using the arrays module is the same efficiency as constructing tuples through other methods.

As the user is not constructing the data structures themselves I think use of the module will only serve to  add a confusing layer of misdirection.

Neither are idiomatic Elixir.

**Comment by martinsvalin** [comment]
Any data conversion, or checking that inputs are sorted is likely to completely blow the O(log n) budget. The only reasonable way to have a binary search implementation that actually achieves O(log n) is to only do the search itself on the inputs.

Now, an idiomatic way of doing binary search in Erlang/Elixir in practice would be to use `:gb_trees` or something similar. There's already a `:gb_trees.lookup/2` function I assume does binary search on the tree. But we don't want people to use the existing `lookup`; we want them to implement and understand binary search, without teaching anti-patterns.

This isn't the main use case of tuples, I agree. But I wouldn't go so far as to say that it's an anti-pattern to use tuples for their constant time random access. They are the building block of many other data structures, including `:arrays` and `:gb_trees`.

**Comment by parkerl** [comment]
Closed by #251
